<img src="https://uploads-ssl.webflow.com/628b99344f25667e77da83cf/62c3a95fc598e35cf796a1f2_Asset%209%403x-p-800.png" alt="drawing" width="400"/>

<center><strong><span style = "font-size:30px"> Chapter 2: Python Fundamentals</span></strong></center>
<br></br>
<center><strong><span style = "font-size:20px"> Project 1 - US Election 2020</span></strong></center>

---

### Reading of csv file

In class, we have manually read tables by defining the lists inside the tables ourselves. But in reality, it is impractical, for instance, `president_county_candidate.csv` has 31407 rows and manually defining every row is too time consuming. Luckily, python has a library `csv` which allows us to read the csv files and convert them to lists of lists e.g. table = [row1, row2, row3, ...., row31407].

Below is how the <i>first 5</i> rows of table `election_data` looks like in its tabular form:

| state    | county            | candidate     | party | total_votes |
|----------|-------------------|---------------|-------|-------------|
| Delaware | Kent County       | Joe Biden     | DEM   | 44552       |
| Delaware | Kent County       | Donald Trump  | REP   | 41009       |
| Delaware | Kent County       | Jo Jorgensen  | LIB   | 1044        |
| Delaware | Kent County       | Howie Hawkins | GRN   | 420         |
| Delaware | New Castle County | Joe Biden     | DEM   | 195034      |

In [5]:
opened_file = open('../Datasets/president_county_candidate.csv')
from csv import reader
read_file = reader(opened_file)
election_data = list(read_file)

Lets take a look at the first 6 rows (including headers)

In [2]:
election_data[:6]

[['state', 'county', 'candidate', 'party', 'total_votes'],
 ['Delaware', 'Kent County', 'Joe Biden', 'DEM', '44552'],
 ['Delaware', 'Kent County', 'Donald Trump', 'REP', '41009'],
 ['Delaware', 'Kent County', 'Jo Jorgensen', 'LIB', '1044'],
 ['Delaware', 'Kent County', 'Howie Hawkins', 'GRN', '420'],
 ['Delaware', 'New Castle County', 'Joe Biden', 'DEM', '195034']]

### What are the data types for each columns?

Remember to exclude the header (row 0)!

In [3]:
col1_type = type(election_data[1][0])
print('Column 1 type:', col1_type)
col2_type = type(election_data[1][1])
print('Column 2 type:', col2_type)
col3_type = type(election_data[1][2])
print('Column 3 type:', col3_type)
col4_type = type(election_data[1][3])
print('Column 4 type:', col4_type)
col5_type = type(election_data[1][4]) 
print('Column 5 type:', col5_type)

Column 1 type: <class 'str'>
Column 2 type: <class 'str'>
Column 3 type: <class 'str'>
Column 4 type: <class 'str'>
Column 5 type: <class 'str'>


### How many total votes are there among the 4 candidates in "Delaware" state, "Kent County" county?

The information we require resides in row 1 to 4. Note: Notice that column 5, total_votes, is of type `str`, we have to do one extra step before adding them together

In [6]:
election_data

[['state', 'county', 'candidate', 'party', 'total_votes'],
 ['Delaware', 'Kent County', 'Joe Biden', 'DEM', '44552'],
 ['Delaware', 'Kent County', 'Donald Trump', 'REP', '41009'],
 ['Delaware', 'Kent County', 'Jo Jorgensen', 'LIB', '1044'],
 ['Delaware', 'Kent County', 'Howie Hawkins', 'GRN', '420'],
 ['Delaware', 'New Castle County', 'Joe Biden', 'DEM', '195034'],
 ['Delaware', 'New Castle County', 'Donald Trump', 'REP', '88364'],
 ['Delaware', 'New Castle County', 'Jo Jorgensen', 'LIB', '2953'],
 ['Delaware', 'New Castle County', 'Howie Hawkins', 'GRN', '1282'],
 ['Delaware', 'Sussex County', 'Donald Trump', 'REP', '71230'],
 ['Delaware', 'Sussex County', 'Joe Biden', 'DEM', '56682'],
 ['Delaware', 'Sussex County', 'Jo Jorgensen', 'LIB', '1003'],
 ['Delaware', 'Sussex County', 'Howie Hawkins', 'GRN', '437'],
 ['District of Columbia', 'District of Columbia', 'Joe Biden', 'DEM', '38037'],
 ['District of Columbia',
  'District of Columbia',
  'Donald Trump',
  'REP',
  '1703'],
 ['Distr

In [4]:
total_kent_county = int(election_data[1][4]) + int(election_data[2][4])\
                    + int(election_data[3][4]) + int(election_data[4][4])
print(total_kent_county)

87025


### Create an if-else statement to determine whether a vote count is high, low, or middle

- High: More than 30k votes
- Middle: 10k to 30k votes inclusive
- Low: Less than 10k Votes

In [5]:
vote_count = 44552

In [6]:
if vote_count > 30000:
    print('High')
elif vote_count > 10000:
    print('Middle')
else:
    print('Low')

High


### Using for loop, find out the sum of ALL the total votes in the entire table

- Again, remember to exclude row 0 (header)
- remember to convert `str` type to `int` for total_vote column

In [7]:
election_data[1:]

[['Delaware', 'Kent County', 'Joe Biden', 'DEM', '44552'],
 ['Delaware', 'Kent County', 'Donald Trump', 'REP', '41009'],
 ['Delaware', 'Kent County', 'Jo Jorgensen', 'LIB', '1044'],
 ['Delaware', 'Kent County', 'Howie Hawkins', 'GRN', '420'],
 ['Delaware', 'New Castle County', 'Joe Biden', 'DEM', '195034'],
 ['Delaware', 'New Castle County', 'Donald Trump', 'REP', '88364'],
 ['Delaware', 'New Castle County', 'Jo Jorgensen', 'LIB', '2953'],
 ['Delaware', 'New Castle County', 'Howie Hawkins', 'GRN', '1282'],
 ['Delaware', 'Sussex County', 'Donald Trump', 'REP', '71230'],
 ['Delaware', 'Sussex County', 'Joe Biden', 'DEM', '56682'],
 ['Delaware', 'Sussex County', 'Jo Jorgensen', 'LIB', '1003'],
 ['Delaware', 'Sussex County', 'Howie Hawkins', 'GRN', '437'],
 ['District of Columbia', 'District of Columbia', 'Joe Biden', 'DEM', '38037'],
 ['District of Columbia',
  'District of Columbia',
  'Donald Trump',
  'REP',
  '1703'],
 ['District of Columbia', 'District of Columbia', ' Write-ins', 'WR

In [14]:
total_table = 0 #define your table first
for row in election_data[1:]:
    total_table += int(row[-1])
print (total_table)

In [7]:
total_table = 0
for row in election_data[1:]:
    total_table += int(row[-1])
print(total_table)

157575102


### Using for loop, create a dictionary contains key:pair value of state:total_votes_in_state

Your resulting dictionary should look like:
    
    {'Delaware': 504010,
     'District of Columbia': 333682,
     'Florida': 11067456,
     'Georgia': 4998566,
     'Hawaii': 574469,
     ...}

In [21]:
state_sum = {}
for row in election_data[1:]:
    if row[0] in state_sum:
        state_sum[row[0]] += int(row[-1])
    else:
        state_sum[row[0]] = int(row[-1])
print(state_sum)

{'Delaware': 504010, 'District of Columbia': 333682, 'Florida': 11067456, 'Georgia': 4998566, 'Hawaii': 574469, 'Idaho': 868105, 'Illinois': 7170200, 'Indiana': 3035088, 'Iowa': 1688310, 'Kansas': 1334330, 'Kentucky': 2136768, 'Louisiana': 2148062, 'Maine': 786952, 'Maryland': 2998114, 'Massachusetts': 3549597, 'Michigan': 5544102, 'Minnesota': 3268708, 'Mississippi': 1315240, 'Missouri': 3028133, 'Montana': 605750, 'Nebraska': 951390, 'Nevada': 1405376, 'New Hampshire': 804430, 'New Jersey': 4551202, 'New Mexico': 923829, 'New York': 7073212, 'North Carolina': 5524813, 'North Dakota': 361819, 'Ohio': 5925806, 'Oklahoma': 1560699, 'Oregon': 2385441, 'Pennsylvania': 6915514, 'Rhode Island': 513602, 'South Carolina': 2513329, 'South Dakota': 422609, 'Tennessee': 3049514, 'Texas': 11323963, 'Utah': 1487696, 'Vermont': 367428, 'Virginia': 4456310, 'Washington': 4053056, 'West Virginia': 794204, 'Wisconsin': 3297434, 'Wyoming': 276765, 'Alabama': 2309900, 'Alaska': 356796, 'Arkansas': 12176

### Using for loop, create a dictionary that contains `state` as `key` and `value` as another dictionary of the vote breakdown by president

Your resulting dictionary should look like:
    
    {'Delaware': 
         {'Joe Biden': 296268,
          'Donald Trump': 200603,
          'Jo Jorgensen': 5000,
          'Howie Hawkins': 2139},
     'District of Columbia': 
         {'Joe Biden': 307265,
          'Donald Trump': 18172,
          ' Write-ins': 3083,
          'Howie Hawkins': 1667,
          'Jo Jorgensen': 1977,
          'Gloria La Riva': 839,
          'Brock Pierce': 679},
      ...}

In [22]:
state_cand_sum = {}
for row in election_data[1:]:
    if row[0] in state_cand_sum:
        inner_dic = state_cand_sum[row[0]]
        if row[2] in inner_dic:
            inner_dic[row[2]] += int(row[-1])
        else:
            inner_dic[row[2]] = int(row[-1])
    else:
        state_cand_sum[row[0]] = {}
        state_cand_sum[row[0]][row[2]] = int(row[-1])
        
print(state_cand_sum)

{'Delaware': {'Joe Biden': 296268, 'Donald Trump': 200603, 'Jo Jorgensen': 5000, 'Howie Hawkins': 2139}, 'District of Columbia': {'Joe Biden': 307265, 'Donald Trump': 18172, ' Write-ins': 3083, 'Howie Hawkins': 1667, 'Jo Jorgensen': 1977, 'Gloria La Riva': 839, 'Brock Pierce': 679}, 'Florida': {'Joe Biden': 5297045, 'Donald Trump': 5668731, 'Jo Jorgensen': 70324, 'Howie Hawkins': 14721, 'Gloria La Riva': 5712, 'Rocky De La Fuente': 5966, 'Don Blankenship': 3902, ' Write-ins': 1055}, 'Georgia': {'Donald Trump': 2461837, 'Joe Biden': 2474507, 'Jo Jorgensen': 62138, ' Write-ins': 84}, 'Hawaii': {'Joe Biden': 366130, 'Donald Trump': 196864, 'Jo Jorgensen': 5539, 'Howie Hawkins': 3822, 'Brock Pierce': 1183, 'Don Blankenship': 931}, 'Idaho': {'Donald Trump': 554118, 'Joe Biden': 287021, 'Jo Jorgensen': 16404, 'Kanye West': 3632, 'Brock Pierce': 2808, 'Don Blankenship': 1886, 'Rocky De La Fuente': 1491, ' Write-ins': 745}, 'Illinois': {'Donald Trump': 2809201, 'Joe Biden': 4223888, 'Jo Jorgen

### Define a function that takes in a candidate name and returns the sum of total_votes the candidates received in the entire table. Using a for loop on the newly created dictionary `state_cand_sum`, create a new dictionary that counts the total number of votes each candidate has

Your resulting dictionary should look like:
    
    {'Joe Biden': 80562197,
     'Donald Trump': 74142138,
     'Jo Jorgensen': 1859226,
     'Howie Hawkins': 394996,
     ' Write-ins': 178640,
     ...}

In [24]:
state_cand_sum

{'Delaware': {'Joe Biden': 296268,
  'Donald Trump': 200603,
  'Jo Jorgensen': 5000,
  'Howie Hawkins': 2139},
 'District of Columbia': {'Joe Biden': 307265,
  'Donald Trump': 18172,
  ' Write-ins': 3083,
  'Howie Hawkins': 1667,
  'Jo Jorgensen': 1977,
  'Gloria La Riva': 839,
  'Brock Pierce': 679},
 'Florida': {'Joe Biden': 5297045,
  'Donald Trump': 5668731,
  'Jo Jorgensen': 70324,
  'Howie Hawkins': 14721,
  'Gloria La Riva': 5712,
  'Rocky De La Fuente': 5966,
  'Don Blankenship': 3902,
  ' Write-ins': 1055},
 'Georgia': {'Donald Trump': 2461837,
  'Joe Biden': 2474507,
  'Jo Jorgensen': 62138,
  ' Write-ins': 84},
 'Hawaii': {'Joe Biden': 366130,
  'Donald Trump': 196864,
  'Jo Jorgensen': 5539,
  'Howie Hawkins': 3822,
  'Brock Pierce': 1183,
  'Don Blankenship': 931},
 'Idaho': {'Donald Trump': 554118,
  'Joe Biden': 287021,
  'Jo Jorgensen': 16404,
  'Kanye West': 3632,
  'Brock Pierce': 2808,
  'Don Blankenship': 1886,
  'Rocky De La Fuente': 1491,
  ' Write-ins': 745},
 'I

In [10]:
cand_sum ={}
for state in state_cand_sum:
    inner_dic = state_cand_sum[state]
    for candidate in inner_dic:
        if candidate in cand_sum:
            cand_sum[candidate] += inner_dic[candidate]
        else:
            cand_sum[candidate] = inner_dic[candidate]

print(cand_sum)

{'Joe Biden': 80562197, 'Donald Trump': 74142138, 'Jo Jorgensen': 1859226, 'Howie Hawkins': 394996, ' Write-ins': 178640, 'Gloria La Riva': 84501, 'Brock Pierce': 46276, 'Rocky De La Fuente': 87220, 'Don Blankenship': 60456, 'Kanye West': 66329, 'Brian Carroll': 25137, 'Ricki Sue King': 547, 'Jade Simmons': 5280, 'President Boddie': 1125, 'Bill Hammons': 6645, 'Tom Hoefling': 668, 'Alyson Kennedy': 6794, 'Jerome Segal': 5854, 'Phil Collins': 4821, ' None of these candidates': 14079, 'Sheila Samm Tittle': 1807, 'Dario Hunter': 5385, 'Joe McHugh': 2843, 'Christopher LaFontaine': 856, 'Keith McCormic': 126, 'Brooke Paige': 1175, 'Gary Swing': 141, 'Richard Duncan': 213, 'Blake Huber': 409, 'Kyle Kopitke': 815, 'Zachary Scalf': 29, 'Jesse Ventura': 2663, 'Connie Gammon': 1473, 'John Richard Myers': 1365, 'Mark Charles': 2007, 'Princess Jacob-Fambro': 495, 'Joseph Kishore': 196, 'Jordan Scott': 175}


NameError: name 'state' is not defined

### Create a function that takes in 2 parameters, `state` and `county`, and return the maximum vote count and the candidate name among all the candidates of the state-county.

`print(max_state_county('Delaware', 'Kent County'))` should return:
       
    (44552, 'Joe Biden')

In [27]:
def max_state_county(state, county):
    state_county_max = 0
    state_county_cand = 'initial'
    for row in election_data[1:]:
        if row[0] == state and row[1] == county and int(row[-1]) > state_county_max:
            state_county_max = int(row[-1])
            state_county_cand = row[2]
    return state_county_max, state_county_cand

print(max_state_county('Delaware', 'Kent County'))

(44552, 'Joe Biden')


### [DIFFICULT] Finding out the winning candidate

In order to determine whether a candidate has won, the candidates must have the majority votes in the state-county. For example:

| state    | county      | candidate     | party | total_votes | won   |
|----------|-------------|---------------|-------|-------------|-------|
| Delaware | Kent County | Joe Biden     | DEM   | 44552       | True  |
| Delaware | Kent County | Donald Trump  | REP   | 41009       | False |
| Delaware | Kent County | Jo Jorgensen  | LIB   | 1044        | False |
| Delaware | Kent County | Howie Hawkins | GRN   | 420         | False |

In the above table, `Delaware-Kent County` has 4 candidates competing with `Joe Biden` having the most number of `total_votes` among the competing 4 candidates. Hence, it is `True` that he won the `Delaware-Kent County` State-County.

#### Using for loop, create a nested dictionary  that counts the max total_votes for each state-county

The nested dictionary dictionary should look like:

    {'Delaware': {'Kent County': 44552,
                  'New Castle County': 195034,
                  'Sussex County': 71230},
                  ...}

In [12]:
state_county_max = {}
for row in election_data[1:]:
    if row[0] in state_county_max:
        state_dic = state_county_max[row[0]]
        if row[1] in state_dic:
                if int(row[-1]) > state_dic[row[1]]:
                    state_dic[row[1]] = int(row[-1])
                else:
                    continue
        else:
            state_dic[row[1]] = int(row[-1])
    else:
        state_county_max[row[0]] = {}
        state_county_max[row[0]][row[1]] = int(row[-1])

print(state_county_max)

{'Delaware': {'Kent County': 44552, 'New Castle County': 195034, 'Sussex County': 71230}, 'District of Columbia': {'District of Columbia': 38037, 'Ward 2': 28185, 'Ward 3': 38436, 'Ward 4': 41604, 'Ward 5': 42453, 'Ward 6': 52357, 'Ward 7': 35861, 'Ward 8': 30332}, 'Florida': {'Alachua County': 89704, 'Baker County': 11911, 'Bay County': 66097, 'Bradford County': 10334, 'Brevard County': 207883, 'Broward County': 618752, 'Calhoun County': 5274, 'Charlotte County': 73243, 'Citrus County': 65352, 'Clay County': 84480, 'Collier County': 128950, 'Columbia County': 23836, 'DeSoto County': 8313, 'Dixie County': 6759, 'Duval County': 252556, 'Escambia County': 96674, 'Flagler County': 43043, 'Franklin County': 4675, 'Gadsden County': 16153, 'Gilchrist County': 7895, 'Glades County': 3782, 'Gulf County': 6113, 'Hamilton County': 3815, 'Hardee County': 6122, 'Hendry County': 7906, 'Hernando County': 70412, 'Highlands County': 34873, 'Hillsborough County': 376367, 'Holmes County': 8080, 'Indian 

#### Using for loop, we find out if the candidate's total_votes in the row is equals to the most votes in the state-county. If yes, we add 1 to the candidate won counter, if no we will skip and go to the next row.

In [13]:
cand_won = {}
for row in election_data[1:]:
    if int(row[-1]) == state_county_max[row[0]][row[1]]:
        if row[2] in cand_won:
            cand_won[row[2]] += 1
        else:
            cand_won[row[2]] = 1
    else:
        continue

In [14]:
cand_won

{'Joe Biden': 1508,
 'Donald Trump': 3259,
 'Brian Carroll': 1,
 'Howie Hawkins': 132,
 'Jo Jorgensen': 132,
 'Gloria La Riva': 1,
 ' Write-ins': 132,
 'Rocky De La Fuente': 122}